### ISSUE TITLE
[MU4 Issue] Fix to Line attribute missing

### ISSUE THREAD

**Issue by JeffRocchio** [musescore/MuseScore#11855]
Unable to find the '**Fix to Line**' attribute in the inspector

**Expected behavior**
That there would be a 'Fix to Line" attribute setting in the Properties dialog OR stems would automatically scale their length to match the visible noteheads. 

**Platform information**
 - OS: Linux; Nightly build appimage of 5/30/2022.

**Additional context**
See on Musescore.org, [Issue #315065, Note Stems Not Scaling with Hidden Note Heads](https://musescore.org/en/node/315065).

NOTE: I had earlier posted this issue to musescore.org site, but in looking around here it seems like it belongs here. See [Issue 332448, Fix to Line attribute missing in v4](https://musescore.org/en/node/332448)].

(I use this fairly often as a work-around for when I need to show only the notes of a melody-line, hiding the other notes in a chord. It is a work-around because when you hide the bottom notes of a chord stack Musescore does not shorten the staff - and in V4 this behaviour has not changed, so I still need the fix-to-line feature.)

**Comment by Tantacrul** [comment]
We chatted about this a little bit and felt that the right way to solve the issues this feature was designed to work around it to properly solve them. Fix to line is a really obscure functionality and its purpose isn't obvious at all. We are bumping fixes to placement issues for harmonics, slash notation and the case presented here to 4.x.

We'll reopen this issue once 4.0 is out.

**Comment by JeffRocchio** [comment]
Not sure it's proper to add a comment to a closed item...but I am anyway I guess : )   For me I certainly like bumping this out for a 'proper solution.' 

**3.x Score Compatibility**: What will happen when I open a 3.x score where Fix to Line has been applied. In the nightly build I tested here such a score opens and retains it's Fixed to Line formatting. So that's nice. I would just note that, in playing around with that, I couldn't find a way to 'reset' it to remove the fix to line assignment for any given note. For me this is all OK. Simply means I can use 4.0 on such scores as long as I don't need to edit the fix to line stuff. Where I need to use Fix to Line I'll need to continue to use 3.x until a proper fix is implemented.

**Re Use in Plugins**: 2nd question, springing from the above, concerns plugin code. Fix to Line appears to be available in the Plugin API (QVariant | fixedLine), tho I have not yet used it in a plugin I have a todo for myself to write one using it to make my life a bit easier. Presumably this would still be available in 4.0 as, presumably, 4.0 isn't changing any of the underlying structuire, just the UI. Yes? This might be an important point depending on the elapsed time to a proper fix for Fix to Line. For me a plugin could be an effective temporary solution enabling me to let go of 3.x even if the proper fix takes awhile to get to.

**Comment by MarcSabatella** [comment]
Indeed, if the existing "Fix to line" settings in 3.x scores were not still honored, that would definitely bump up the significance of this in my eyes at least.  But that much does work.  More importantly to me, so do the built-in commands that normally set this property - "Tools / Fill with slashes" and "Toggle rhythmic slash notation".  So the property *is* still supported; it just isn't currently exposed to users, unless/until someone creates a plugin.

**Comment by Jojo-Schmitz** [comment]
Time to reopen?

**Comment by aige02** [comment]
Using last night's 4.0.2 nightly (C229b14), I copies slashes (created with Tools=>fill with slashes) from guitar (treble-8vb) to bass (bass-8vb).  While most of the slashes were fixed to position 4, some of the slashes, which had been previously copied from elsewhere in the guitar staff (slashes created using an earlier build), were "transposed" to the new staff.  This may indicate a problem in 4.0.2 with handling the "Fix to Line" flag.  The "Fix to Line" attribute (flag and y-position) would make fixing this cleaner.

Adjusting the Fix-to line is often useful.  Two examples are:
1. A drumset part where the voice 2 is written, but the voice 1 part is ad lib - the slashes need to be moved up.
2. An ad lib solo with a suggested solo where the slashes are usually moved below the suggested solo.

**Comment by GrahamDickins** [comment]
Just to add that I also use this feature and hope to see it in MuseScore 4 :)

Here's how I use it (see screenshot) -- for 'courtesy lyrics' (fitting a sentence on one line by putting words in brackets to indicate it belongs to the previous system), where I attach the lyrics to an invisible Voice 2 note, which I put all at the same note height (line 7) to avoid lyrics being pushed down by auto-place (I have a strict 'one-page score' rule, so line height is a premium!)
![fixtoline](https://user-images.githubusercontent.com/9395025/215548701-b5adbe35-46ac-4d5a-a3d4-1f8a0e2321c7.jpg)

**Comment by MarcSabatella** [comment]
I too think the simplest solution is to just reinstate the option rather than trying to come up with a dozen different solutions for the dozen different use cases.

However, that said, I do think a number of the use cases being discussed here can be solved more effectively by different means - including means already available.  For instance, to force slash notation onto a different staff line, a simple vertical offset seems more consistent with how other similar adjustments would be made.  And regarding the invisible note for lyrics, invisible notes shouldn't affect lyric position at all, so the fact that they do is just a bug, and a new issue should be opened for it.

**Comment by JeffRocchio** [comment]
Plea: Can this simply be added back into the user interface for compatibility with MS-3? I can't use MS-4 without this feature and the 'correct' fix isn't likely to appear in my lifetime. Which is perfectly understandable. But could we at least have back what was in MS-3? Please? Pretty-please?

**Comment by MarcSabatella** [comment]
It’s certainly possible, but have you tried the plugin? It accomplishes the same thing.

Also, it would help greatly if you explained more about your unusual special use case that causes you to not be able to use the program without some sort of functionality that replaces what this feature did.  Then we can better consider how that use case would be handled better without necessarily needing to resort to that specific solution, which was really never more than a hack.

As mentioned above, I’d still like to see the feature back too, but even more so, I’d like to see better solutions to the real world problems people were using this feature for.

**Comment by JeffRocchio** [comment]
> but have you tried the plugin? It accomplishes the same thing.

Somebody made a plugin for this? I missed that. I'll go look for it.
 
> Also, it would help greatly if you explained more about your unusual special use case

For Mountain Dulcimer TAB the standard convention is to also include a SMN staff above the TAB that contains only the notes of the melody line - not the full chord. But I really do need to have all the SMN notes existing on the staff, for a few reasons, including how I coerce Musescore into producing the odd Mountain Dulcimer TAB. So I need to hide the non-melody notes of each chord. In TablEdit you can simply "hide" a note and it properly resizes the staff part to fit only the visible notes. Which is what I would say is the proper way to handle a hidden note. Since Musescore doesn't resize the staff to match the visible noteheads I use the fix-to-line feature to retain the notes but make it look like they are hidden (that is, I fix the 'hidden' chord's notes to the line the melody note is on so that they are hidden behind the melody note.

It's a kludge. But...what else can I do...other than try to use TablEdit on Linux under Wine...which would be a bit of misery.

So I may be the only person on the planet with this use-case. Which, of course, should mean it's not a high priority for the developers.

**Comment by MarcSabatella** [comment]
Yes, there is a fix to line plugin see the main plugin download page at musescore.org.

I'm not understanding in what sense merely making the notes invisible doesn't work - why a staff would ever need to be "resized" in this context.  Or for that matter why not simply not add those notes to the tab staff rather than making them invisible.  Best to ask for help in the official Supprot forum at musescore.org and attach an actual sample score there.  Probably there is a far simpler way to do whatever it is you doing, but we'd need to see a clear example of the desired result to advise better.

**Comment by JeffRocchio** [comment]
> why a staff would ever need to be "resized" in this context.

Whoops - by 'staff' I really meant 'stem.'  I did just edit my last posting to correct that wording.

> Best to ask for help in the official Supprot forum at musescore.org ... need to see a clear example of the desired result

Yes. I think I may have done that actually a long time ago. I'll check and perhaps post something. I am fairly certain that Fix-to-Line is the only practical answer give Musescore's current feature set. But I'd be worth validating that.

**Comment by JeffRocchio** [comment]
> I'm not understanding in what sense merely making the notes invisible doesn't work...Best to ask for help in the official Supprot forum at musescore.org and attach an actual sample score there. Probably there is a far simpler way to do whatever it is you doing, but we'd need to see a clear example of the desired result to advise better.

Marc - and any other interested party - please see: [Fix to Line - Use Case For Mountain Dulcimer TAB](https://musescore.org/en/node/368085)

**Comment by davidstephengrant** [comment]
Reading through this it seems that the specific use-cases described are achievable by using a different (existing and supported) approach in MSS. I'm going to close this, but please let us know if I've overlooked anything for more general cases.

@avvvvve FYI.

**Comment by MarcSabatella** [comment]
No, there are still valid use cases that now require hacks that don't work as well - mostly on account of ledger lines.

The original use case for fix to line was for slash notation, to set the vertical position for the slashes.  For this use case, when you wish to override the default, it works to use a vertical offset, since you seldom if ever want ledger lines on slash notation.

But people have also used this feature to handle cases where you want a note to appear on one staff line but plays back using a different pitch.  This can come up when using harmonics or particular organ stops or other cases where the playback sound for a given pitch is other than what is actually notated.  In some cases, of course, the sound library uses the appropriate pitch for the notation, but with all the different notations and sound libraries that exist, there will always be cases where you want to see a note on one line but hear it played back on another.  Offset won't do the job here if ledger lines are involved, because they won't appear or disappear according to the adjusted vertical position - they are tied to the actual staff line for the pitch.

So for example, here is an attempt to get the playback of the harmonic to use the pitch an octave higher and applying an offset:

<img width="417" height="257" alt="Image" src="https://github.com/user-attachments/assets/89891b07-aa8e-4465-b999-086a491326a4" />

Left looks as desired but depending on the instrument and sound library, might not sound correct.  Right sounds correct but now you see the ledger lines, which moved with the note while adjusting.  Other adjustments might leave the ledger line in their original places, which is just as bad.  Or might leave you with a note floating high above the staff with no ledger lines.  Also stem lengths may be wrong, as in the above example (those can be adjusted, at least).

Workarounds then can be to use notes in other voices, but that often introduces more artifacts to then need to work around.

The fix to line feature works just fine internally; it's just that it was removed from the Properties panel unnecessarily and would be quite simple to just add back.

**Comment by davidstephengrant** [comment]
@MarcSabatella thank you very much for the additional context.

@avvvvve Please, can you triage? Thanks!

**Comment by avvvvve** [comment]
> Workarounds then can be to use notes in other voices, but that often introduces more artifacts to then need to work around.

This workaround, to me at least, is still much more intuitive than the 'Fix to line' feature. Once you move a note with 'Fix to line', there's no visual indication in the score that the note has been visually moved from where it actually is/how it will play back.

So, we could reintroduce something _like_ 'Fix to line' someday, but should really be renamed for clarity (i.e. something to the effect of 'Decouple drawn note from played note'), let you specify the pitch as in 'A4' rather than calling the top line 0, and perhaps show a ghost note where the played note remains.

**Comment by MarcSabatella** [comment]
If you *know* about fix to line, it’s definitely *easier* than playing with hidden voices for this, but indeed, probably not as *intuitive*.  I would say the bigger downside of using fix to line for the harmonics case is that by fixing to a staff line, it’s wrong if you then transpose the passage or change clefs.  A more direct way to have a pitch that plays back as another pitch would be much better for that case.  So fix to line is what people had been resorting to but is far from ideal.

The more appropriate use cases for fix to line specifically are more, “things like slash notation”, where you specifically *want* the note to be impervious to transposition or clef changes.  The actual implementation of the slash notation feature also suppressed accidentals for the note, BTW.

To me the question is, aside from the actual slash notation feature, what *other* “things like slash notation” are there - use cases there that truly want to fix to the staff line (and thus independent of transposition or clef changes) rather than to pitch.  I can still think of a couple:

* spoken word passages in vocal lines
* notation of a single unpitched percussion note on a staff otherwise being used for a pitched instrument (e.g., a marimba player hitting an attached cowbell or triangle, or a pianist asked to slap the desk, or trumpet player doing a finger snap)

In these cases, it truly is the staff line you want to specify, and you want accidentals suppressed, but you don’t necessary want a slash head - a cross might be more appropriate.  Luckily, out does basically work to use rhythmic slash notation and then change the notehead and apply an offset if desired.  But I’d argue this also is not so intuitive.

So if a “decouple drawn note” feature is added, maybe it could handle *both* the actual pitched case *and* these cases?  Our rename and expand the functionality of the existing rhythmic slash notation command.
